src/meta/src/barrier/progress.rs

```diff
@@ -0,0 +1,150 @@
+        }
+    }
+
+    /// Update the progress of `actor` according to the given epochs.
+    fn update(&mut self, actor: ActorId, consumed_epoch: Epoch, current_epoch: Epoch) {
+        match self.states.get_mut(&actor).unwrap() {
+            state @ (ChainState::ConsumingSnapshot | ChainState::ConsumingUpstream(_)) => {
+                if consumed_epoch == current_epoch {
+                    *state = ChainState::Done;
+                    self.done_count += 1;
+                } else {
+                    *state = ChainState::ConsumingUpstream(consumed_epoch);
+                }
+            }
+            ChainState::Done => panic!("should not report progress after done"),
+        }
+    }
+
+    /// Returns whether all chains are done.
+    fn is_done(&self) -> bool {
+        self.done_count == self.states.len()
+    }
+
+    /// Returns the ids of all actors containing the chain nodes for the mview tracked by this
+    /// [`Progress`].
+    fn actors(&self) -> impl Iterator<Item = ActorId> + '_ {
+        self.states.keys().cloned()
+    }
+}
+
+/// Track the progress of all creating mviews. When creation is done, `notify_finished` will be
+/// called on registered notifiers.
+#[derive(Default)]
+pub(super) struct CreateMviewProgressTracker {
+    /// Progress of the create-mview DDL indicated by the epoch.
+    progress_map: HashMap<CreateMviewEpoch, (Progress, Vec<Notifier>)>,
+
+    /// Find the epoch of the create-mview DDL by the actor containing the chain node.
+    actor_map: HashMap<ActorId, CreateMviewEpoch>,
+}
+
+impl CreateMviewProgressTracker {
+    /// Add a new create-mview DDL command to track with current epoch as `ddl_epoch` and
+    /// `notifiers`, that needs to wait for `actors` to report progress.
+    ///
+    /// If `actors` is empty, [`Notifier::notify_finished`] will be called immediately.
+    pub fn add(
+        &mut self,
+        ddl_epoch: Epoch,
+        actors: impl IntoIterator<Item = ActorId>,
+        notifiers: impl IntoIterator<Item = Notifier>,
+    ) {
+        let actors = actors.into_iter().collect_vec();
+        if actors.is_empty() {
+            // The command can be finished immediately.
+            notifiers.into_iter().for_each(Notifier::notify_finished);
+            return;
+        }
+
+        for &actor in &actors {
+            self.actor_map.insert(actor, ddl_epoch);
+        }
+
+        let progress = Progress::new(actors);
+        let notifiers = notifiers.into_iter().collect();
+        let old = self.progress_map.insert(ddl_epoch, (progress, notifiers));
+        assert!(old.is_none());
+    }
+
+    /// Update the progress of `actor` according to the given epochs. If all actors in this MV have
+    /// finished, `notify_finished` will be called on registered notifiers.
+    pub fn update(&mut self, actor: ActorId, consumed_epoch: Epoch, current_epoch: Epoch) {
+        let epoch = self.actor_map.get(&actor).cloned().unwrap_or_else(|| {
+            panic!(
```

Comment:
Even after recovery, update will never receives bad actors. Should be `unreachable!()` I think.

Reply:
I decide to ignore the not-found case here since some finishing cases of the rearranged chain may lead to a progress update after finished. Will investigate it in the next PR for reporting progress.